util: Update & fix bug in m5stats2streamline.py

atrah22 opened the pull request:
There is conversion error in ./util/streamline/m5stats2streamline.py script to convert gem5 stats.txt,sys, system.tasks.txt to the apc folder required by DS-5 streamline. The fix to the bug can convert to apc folder without error. The zipped apc folder can then be imported in older DS-5 v5.24 for visualization (didn't work with DS-5 v5.29).

Changes:
1) writeBinary function binary_list can have either string or ints and it needs to be properly converted to bytes 
2) packed32(x) function can have x as int or float. Incase of float it needs to be converted to int

The bug was reported and solved primarily in the issue https://github.com/gem5/gem5/issues/145 

Change-Id: I6a52aa59e1582dd6bb06b2d1c49ddaf8fe61c997

Commit:
util: Update & fix bug in m5stats2streamline.py

1) writeBinary function binary_list can have either string or ints and it needs to be properly converted to bytes
2) packed32(x) function can have x as int or float. Incase of float it needs to be converted to int

Change-Id: I6a52aa59e1582dd6bb06b2d1c49ddaf8fe61c997

giactra commented on util/streamline/m5stats2streamline.py:
```diff
@@ -661,7 +668,7 @@ def parseProcessInfo(task_file):
 
     for line in process_file:
-        match = re.match(process_re, line)
+        match = re.match(process_re, line.decode("ISO-8859-1"))
```
Any reason to use "ISO-8859-1" over UTF-8?

atrah22 replied:
> Any reason to use "ISO-8859-1" over UTF-8? 
--> no, there ain't a solid reason.
Actually, using decode() would resolve the issue anyway and is better approach as you mentioned.

giactra commented:
Thanks for the PR! LGTM, there's only one thing to clarify (see encoding comment)

atrah22 commented:
> Thanks for the PR! LGTM, there's only one thing to clarify (see encoding comment)

Hello, thanks.
The  line match = re.match(process_re, line) gives the following error (python 3.x) "TypeError: cannot use a string pattern on a bytes-like object"
So I explicitly encoded the line to string of characters and changed hence the line to match = re.match(process_re, line.decode('ISO-8859-1'))

giactra commented:
IMO we should use UTF-8 which is a superset of ISO-8859-1, so probably doing something like  re.match(process_re, line.decode()) will still solve your python problem

giactra commented:
Another thing I just noticed is that you are trying to merge this PR with the stable branch... Our typical workflow is to merge to develop instead so I might ask you to move/open a new PR (apologies) targeting gem5:develop

atrah22 commented:
> Another thing I just noticed is that you are trying to merge this PR with the stable branch... Our typical workflow is to merge to develop instead so I might ask you to move/open a new PR (apologies) targeting gem5:develop

Ah okay, I can do so (new PR in gem5:develop). 
I misunderstood, as I saw also other PR requests on the main stable branch.
I will then first make the change and move/open this/new PR in gem5:develop

Commit:
util: Update & fix bug in m5stats2streamline.py

1) writeBinary function binary_list can have either string or ints and it needs to be properly converted to bytes
2) packed32(x) function can have x as int or float. Incase of float it needs to be converted to int
3) encode lines to string using .decode() or else TypeError will be invoked during run

Change-Id: I678169f191901f02a80187418a17adbc1240c7d3

BobbyRBruce commented:
@atrah22, it seems you're having some issues here. I believe you applied these changes on top of the stable branch, and now that you're trying to merge these changes into develop and git has got confused. The fix is actually quite simple. Do the following in your repository:

```sh
# I'm assuming the branch with these changes is 'streamline'.
# Let's start there.
git switch streamline

# Add the gem5 mainline as a remote repo (you may have
# already done this)
git remote add upstream git@github.com:gem5/gem5.git

# Obtain the upstream (gem5 mainline) develop branch.
# Let's call it 'upstream-develop'.
git switch -c upstream-develop --track upstream/develop

# Create a new branch from 'upstream-develop'. Let's call this
'streamline-new'
git switch -c streamline-new

# Cherry-pick your changes onto this new branch.
git cherry-pick ee08cefdaa0b03feb3f5bb120338b6417c66c81d
git cherry-pick deaf37173d0befbb4afaf6533482ecede6fb69c3

# Rename your old streamline branch 'streamline-old'
git branch -M streamline streamline-old

# Rename the current 'streamline-new' to 'streamline'
git branch -M streamline-new streamline

# Now we've cleaned everything up. All you need to do is 'force push'.
git push -f
```

Once you do that this PR should be looking a lot better.

atrah22 commented:
> @atrah22, it seems you're having some issues here. I believe you applied these changes on top of the stable branch, and now that you're trying to merge these changes into develop and git has got confused. The fix is actually quite simple. Do the following in your repository:
> 
> ```shell
> # I'm assuming the branch with these changes is 'streamline'.
> # Let's start there.
> git switch streamline
> 
> # Add the gem5 mainline as a remote repo (you may have
> # already done this)
> git remote add upstream git@github.com:gem5/gem5.git
> 
> # Obtain the upstream (gem5 mainline) develop branch.
> # Let's call it 'upstream-develop'.
> git switch -c upstream-develop --track upstream/develop
> 
> # Create a new branch from 'upstream-develop'. Let's call this
> 'streamline-new'
> git switch -c streamline-new
> 
> # Cherry-pick your changes onto this new branch.
> git cherry-pick ee08cefdaa0b03feb3f5bb120338b6417c66c81d
> git cherry-pick deaf37173d0befbb4afaf6533482ecede6fb69c3
> 
> # Rename your old streamline branch 'streamline-old'
> git branch -M streamline streamline-old
> 
> # Rename the current 'streamline-new' to 'streamline'
> git branch -M streamline-new streamline
> 
> # Now we've cleaned everything up. All you need to do is 'force push'.
> git push -f
> ```
> 
> Once you do that this PR should be looking a lot better.

Thanks for the help. I merged the develop branch and used git cherry-pick to add the previous commits. Hopefully now there is no merge conflict.

BobbyRBruce reviewed (approved):
This makes sense to me but as @giactra is the assignee i'll wait for him to double check and merge.

giactra reviewed (approved):
LGTM, thanks for the patience!
